crates/noirc_evaluator/src/ssa_refactor/ir/dom.rs

```diff
@@ -0,0 +1,426 @@
+        &self,
+        block_id: BasicBlockId,
+        cfg: &ControlFlowGraph,
+    ) -> BasicBlockId {
+        // Get an iterator with just the reachable, already visited predecessors to `block_id`.
+        // Note that during the first pass `node` was pre-populated with all reachable blocks.
+        let mut reachable_preds =
+            cfg.pred_iter(block_id).filter(|pred_id| self.nodes.contains_key(pred_id));
+
+        // This function isn't called on unreachable blocks or the entry block, so the reverse
+        // post-order will contain at least one predecessor to this block.
+        let mut immediate_dominator =
+            reachable_preds.next().expect("block node must have one reachable predecessor");
+
+        for pred in reachable_preds {
+            immediate_dominator = self.common_dominator(immediate_dominator, pred);
+        }
+
+        immediate_dominator
+    }
+
+    /// Compute the common dominator of two basic blocks.
+    ///
+    /// Both basic blocks are assumed to be reachable.
+    fn common_dominator(
+        &self,
+        mut block_a_id: BasicBlockId,
+        mut block_b_id: BasicBlockId,
+    ) -> BasicBlockId {
+        loop {
+            match self.reverse_post_order_cmp(block_a_id, block_b_id) {
+                Ordering::Less => {
+                    // "a" comes before "b" in the reverse post-order. Move "b" up.
+                    block_b_id = self.nodes[&block_b_id]
+                        .immediate_dominator
+                        .expect("Unreachable basic block?");
+                }
+                Ordering::Greater => {
+                    // "b" comes before "a" in the reverse post-order. Move "a" up.
+                    block_a_id = self.nodes[&block_a_id]
+                        .immediate_dominator
+                        .expect("Unreachable basic block?");
+                }
+                Ordering::Equal => break,
+            }
+        }
+
+        debug_assert_eq!(block_a_id, block_b_id, "Unreachable block passed to common_dominator?");
+        block_a_id
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::cmp::Ordering;
+
+    use crate::ssa_refactor::{
+        ir::{
+            basic_block::BasicBlockId, cfg::ControlFlowGraph, dom::DominatorTree,
+            function::Function, instruction::TerminatorInstruction, map::Id, post_order::PostOrder,
+            types::Type,
+        },
+        ssa_builder::FunctionBuilder,
+    };
+
+    #[test]
+    fn empty() {
+        let func_id = Id::test_new(0);
+        let mut func = Function::new("func".into(), func_id);
+        let block0_id = func.entry_block();
+        func.dfg.set_block_terminator(
+            block0_id,
+            TerminatorInstruction::Return { return_values: vec![] },
+        );
+        let cfg = ControlFlowGraph::with_function(&func);
+        let post_order = PostOrder::with_function(&func);
+        let dom_tree = DominatorTree::with_cfg_and_post_order(&cfg, &post_order);
```

Comment:
Seems like this is a common operation. Are we expected to keep around the cfg and post_order blocks without the dominator tree? If not we can have a helper function to create the first two and pass them into the third instead of creating all three separately.

Reply:
I thought I'd come across a use case of post order independent of dom trees - but possibly I imagined it. What cranelift does is to have the dom tree compute and take ownership of the post order. It then has an accessor for when the two are used in conjunction. Would that be preferred?

Either way I'll add a helper for the cfg.

Reply:
Not my imagination - cranelift's dead code elimination pass iterates over post-order without any need for dom trees.